src/components/hooks/useActivityStream.js

```diff
@@ -1,53 +1,129 @@
 
+const PREPEND_ACTIVITIES = 'prepend_activities';
 const APPEND_ACTIVITIES = 'append_activities';
 
 /**
- * Returns a new array of activityIDs based on the given action.
- * If no action is passed, it will return the same activityIDs
+ * Returns a new array of activities based on the given action.
+ * If no action is passed, it will return the same activities
  * array without any changes.
  *
- * @param {Array.<string>} activityIDs activityIDs associated to the room
- * @param {object} action action to apply to given activityIDs
- * @returns {Array.<string>} New activityIDs associated to the room
+ * @param {Array.<string|ActivityDate>} activities activities associated to the room
+ * @param {object} action action to apply to given activities
+ * @returns {Array.<string|ActivityDate>}
  */
-function reducer(activityIDs, action) {
-  let newActivityIDs = [];
+function reducer(activities, action) {
+  let newActivities = [];
 
   switch (action.type) {
+    case PREPEND_ACTIVITIES:
+      newActivities = action.payload.reverse().concat(activities);
+      break;
     case APPEND_ACTIVITIES:
-      newActivityIDs = activityIDs.concat(action.payload);
+      newActivities = activities.concat(action.payload);
       break;
     default:
-      newActivityIDs = activityIDs;
+      newActivities = activities;
       break;
   }
 
-  return newActivityIDs;
+  return newActivities;
 }
 
 /**
  * Custom hook that returns activity data associated to the room of the given ID.
+ * If an element reference is given, it will attach an scroll listener to load more
+ * data as user scrolls to top.
  *
  * @param {string} roomID  ID of the room for which to return data.
- * @returns {Room} Activity ID associated to the room
+ * @param {object} elementRef  reference to the element to attach scroll listener.
+ * @param {object} lastElementRef  reference to the last element within elementRef
+ * @returns {Array} Activities state and showLoader state
  */
-export default function useActivityStream(roomID) {
-  const [activityIDs, dispatch] = useReducer(reducer, []);
+export default function useActivityStream(roomID, elementRef, lastElementRef) {
+  const [activities, dispatch] = useReducer(reducer, []);
+  const [showLoader, setShowLoader] = useState(false);
+  const [scrollToLastActivity, setScrollToLastActivity] = useState(true);
   const {roomsAdapter} = useContext(AdapterContext);
 
+  const loadPreviousActivities = async () => {
+    const previousActivities = await roomsAdapter.getPreviousRoomActivities(roomID);
+
+    dispatch({type: PREPEND_ACTIVITIES, payload: previousActivities});
+  };
+
+  const handleScroll = (event) => {
+    // Only load more data when user has reached the top
+    const isViewportTop = event.target.scrollTop === 0;
+
+    if (!showLoader && isViewportTop && roomsAdapter.hasMoreActivities(roomID)) {
+      setShowLoader(true);
+
+      // Throttle scroll
+      setTimeout(async () => {
```

Comment:
is the `setTimeout` still valid here? I don't think we would need it anymore, since we are awaiting.

Reply:
This is how we "fake wait". Data loads too fast and you can't see the loader. 
In general, I think it makes the UX a little better anyway, since we indicate to the user that we actually doing something, even when data comes through really fast. It's more for acknowledgment